### DISCUSSION TITLE
Help with integrating theme-ui and material-ui

### DISCUSSION THREAD

**Discussion by TheFirstMe** [system-ui/theme-ui#819] [Q&A]
( Related to #164 )

I have a Gatsby application that runs on theme-ui. I really love the API that comes with theme-ui. But at the same time I also want to use material-ui components inside my application. Can someone help me on how I would approach this problem? 

In the end, I should be able to access theme-ui theme from the sx prop and it should be passed down to the material-ui components. I’m thinking of a transformer as mentioned in the above issue. Right now, I am using Gatsby plugins for theme-ui and material-ui but it requires two separate theme objects. If someone can give me ideas on how to transform theme-ui theme to material-ui theme, I could work on a PR

**Comment by clarkkozak** [comment]
Here's something that's helped me: https://theme-ui.com/use-theme-ui/

**Comment by TheFirstMe** [comment]
@clarkkozak Sounds good but I'll have to create a different theme object for different use cases. I was thinking of a transformer as mentioned in #164

**Comment by TheFirstMe** [comment]
@jxnblk Do you have any ideas? I was thinking of finding properties of material-ui theme object from the theme-ui one and do the corresponding mapping. What do you think?

**Comment by TheFirstMe** [comment]
material-ui also has a [createMuiTheme](https://material-ui.com/customization/theming/#createmuitheme-options-args-theme) function that creates their theme object. I think this can also be used.

**Comment by clarkkozak** [comment]
@TheFirstMe sounds perfect! Honestly though, I'd close this issue ASAP because it's more asking for help with coding rather than an issue with the code base. I suggest checking out a forum specifically centered around helping others code. For example: https://stackoverflow.com/

**Comment by TheFirstMe** [comment]
@clarkkozak That was my initial intention. But I'd also like to create a feature that does this for theme-ui if possible.

**Comment by clarkkozak** [comment]
@TheFirstMe right on :) well, if you make a fork, let me know and I'd love to help.

**Comment by TheFirstMe** [comment]
@clarkkozak I'll let you know. Thanks :)

**Comment by jxnblk** [comment]
If my understanding is correct, it should be possible to store your theme values canonically in Theme UI's format/schema, but use a transformer to accommodate Material UI's schema. I can see value in adding something like this as an optional package here in Theme UI, if you're interested in helping out.

I'd imagine the API could look something like the following:

```jsx
import { jsx, ThemeProvider } from 'theme-ui'
import { MaterialUIProvider } from '@theme-ui/material-ui'
import theme from './theme'

export default props =>
  <ThemeProvider theme={theme}>
    <MaterialUIProvider>
      {props.children}
    </MaterialUIProvider>
  </ThemeProvider>
```

```

**Comment by TheFirstMe** [comment]
@jxnblk One thing I found confusing from Theme UI's docs is that not all values are required in the schema but that's not the case in Material UI. I tried building a mock transformer using the createMuiTheme from Material UI's API and it works fine (not all values from Theme UI theme were used). 

Should the transformer account for all possible values in Theme UI's schema that are relevant to Material UI's schema? There are also values in Material UI schema that are not relevant to Theme UI so those should be left untouched?

**Comment by jxnblk** [comment]
If Material UI requires all fields in its API, you could provide those as defaults (I'm surprised that library doesn't do that already 🤔)

**Comment by TheFirstMe** [comment]
I am not sure if they are absolutely required but they don't state that it's not in the docs unlike Theme UI. Also their default theme is generated using their createMuiTheme API. I suppose those fields are necessary because all components are styled based on those values.

**Comment by TheFirstMe** [comment]
My bad I just saw they are using the theme generated from that function by default so it's fine.

**Comment by TheFirstMe** [comment]
As I said above, there are fields used by Material UI that are not used by Theme UI. What should we do about them?

Here's the default theme used by Material UI:

```
{
   breakpoints: {
     keys: ["xs", "sm", "md", "lg", "xl"],
     values: {
       xs: 0,
       ... //remaining breakpoints
     },
     ..... //functions for media queries
   },
   mixins: {
     .... //for gutters and toolbar
   },
   palette: {
     common: {
       black: ..., //value for color
       white: ...,
     }
     type: "light", //light mode
     primary: {
       light: ...,
       main: ..., //only this field is required
       dark: ...,
       contrastText: ...,
     },
     ... //other intentions like secondary, error, success
     grey: {
       ... //different shades of grey
     },
     ...// functions to calculate other fields from intention.main
     //The rest of the palette properites are for light and dark mode
     text: {
       primary: ...,
       ...
     },
     divider: ...,
     background: {
       default: ...,
       paper: ...,
     },
     action: {
       active: ...,
       hover: ...,
       focus: ...,
       ...
     }
   },
   shadows: [...], //array of values for shadow
   typography: {
     htmlFontSize: 16,
     fontFamily: ..., //value for font-family
     fontWeightLight: 300,
     fontWeightRegular: 400,
     fontWeightMedium: 500,
     fontWeightBold: 700,
     //Below are variants h1,h2, button, caption, subtitle, body
     h1: {
       fontFamily: ... //calculated from above
       fontWeight: ...
       fontSize: ...
       lineHeight: ...
       letterSpacing: ...
     },
     ...
   },
   spacing: f(), // a function to create spacing. Can be replaced
                // by array of values like Theme UI
   shape: {
     borderRadius: 4
   },
   transitions: {
     ... //transition values
   },
   zIndex: {
     ... //z-index values for various components
   }
}
```

**Comment by jxnblk** [comment]
The only thing that looks like something not in Theme UI is the `mixins` (which I assume is some sort of functional values??). Was there anything specific that you're unsure of how to map?

**Comment by TheFirstMe** [comment]
I excluded some fields for the sake of simplicity. Here's the full [object](https://material-ui.com/customization/default-theme/#explore). 

In Material UI, the breakpoint keys are fixed for their components and cannot be changed. So breakpoints field in Theme UI should always be an array of 5 values.

In typography variants of Material UI, there are fields from h1 to h6 where only the font-family is the same. The remaining values are calculated by their API. There are also fields like body1, body2, subtitle1, subtitle2. Should we include the same fields in Theme UI as part of the mapping?

I haven't seen the transitions being used in Theme UI examples. So I am also unsure on how to map that. 

And finally fields like typography.actions and typography.divider... should they be accept the same format as in Material UI? I haven't seen those in Theme UI as well.

So basically there are some fields in Material UI not in Theme UI. What should we do about those fields?

**Comment by clarkkozak** [comment]
@TheFirstMe I'm unsure if I understand currently, yet that's where I would suggest using `useThemeUI` to then create your own values for those fields.

If I miss understand, could you give me an example of how you would like to see Material UI fields and Theme UI?

**Comment by TheFirstMe** [comment]
@clarkkozak So its ok to use custom values in Theme UI to match the fields in Material UI? This is the issue that was bugging me. 

I'll give you an example in a bit.

**Comment by TheFirstMe** [comment]
@clarkkozak  Here's the Theme UI theme:

```
{
  "breakpoints": [0, 600, 960, 1280, 1920],
  "space": [0, 4, 8, 16, 32, 64, 128, 256, 512],
  "fonts": {
    "body": '"Roboto", "Helvetica", "Arial", sans-serif',
    "heading": '"Roboto", "Helvetica", "Arial", sans-serif'
  },
  "fontSizes": [12, 14, 16, 20, 24, 32, 48, 64, 96],
  "fontWeights": {
    "body": 400,
    "heading": 700
  },
  "lineHeights": {
    "body": 1.5,
    "heading": 1.125
  },
  "colors": {
    "black": "#000",
    "white": "#fff",

    "primary": "#3f51b5",
    "secondary": "#f50057",
    "error": "#f44336",
    "warning": "#ff9800",
    "info": "#2196f3",
    "success": "#4caf50",

    "text": "rgba(0, 0, 0, 0.87)",
    "background": "#fafafa",

    "modes": {
      "dark": {
        "text": "#fff",
        "background": "#303030",
      }
    }
  },
  "zIndices": {
    "mobileStepper": 1000,
    "speedDial": 1050,
    "appBar": 1100,
    "drawer": 1200,
    "modal": 1300,
    "snackbar": 1400,
    "tooltip": 1500
  }
}
```

The default Material UI theme can be found [here](https://material-ui.com/customization/default-theme/#explore)

The problems: 

- `fonts.body` should be mapped to `typography.fontFamily` or/and `typography.body1.fontFamily` and `typography.body2.fontFamily`
- Theme UI supports both boxShadow and textShadow but Material UI only supports boxShadow in their theme. So the shadows theme key can't be used?
 - There are some fields like mixins, transitions in Material UI which are not in Theme UI. Should we expose those fields throughout the Theme UI theme or somehow pass it only to Material UI theme?

**Comment by clarkkozak** [comment]
@TheFirstMe I see! I'll think about this over my weekend and see if I can think about what it would be like

EDIT: 

Yet I don't think I'll be much help :P you have a great understanding of the problem, and I agree with @jxnblk. Something like `<MaterialUIProvider>` would be awesome to work with.

**Comment by TheFirstMe** [comment]
@clarkkozak I do have an idea on how `<MaterialUIProvider>` will look like. Material UI has their own [Theme Provider](https://material-ui.com/customization/theming/#theme-provider) which accepts a theme object like Theme UI. All we need is to figure out which values to transform to Material UI.

If we can do that, then all we need to do is pass that theme to their ThemeProvider?

EDIT: 
There are slight differences between these two like `colors.text` used as a string value in Theme UI and `palette.text` is an object that has keys like primary and secondary. If its ok to use an object as a value for `colors.text` then things will get more simpler. @jxnblk Thoughts on this? 

EDIT v2:
I saw that we can also use `styles.root` to style the body element. Then that'd solve the above issue. But since it's not mentioned in the docs that we can use objects values for `colors.text` and `colors.background`, I don't know if it's a violation of Theme UI spec or not.

EDIT v3: 
I am not going to touch the colors key for now until I get an update. I'll work with the other keys.

**Comment by TheFirstMe** [comment]
I made a transformer that transforms almost all Theme UI fields to the Material UI with some additional fields that were not in the standard Theme UI object. But I left out some Material UI only fields for now because I need some input from others before I can continue on that. Can I start working on the package now?

**Comment by clarkkozak** [comment]
@TheFirstMe do you have a fork that we can look at? I'd love to see what you got!

**Comment by TheFirstMe** [comment]
@clarkkozak Here's the [fork](https://github.com/TheFirstMe/theme-ui/blob/package-material-ui/packages/material-ui/src/to-mui.js). But I am having trouble running it as a package. I suspect it's because I'm using `@material-ui/core` but haven't included it as a dependency (it's currently a peer dependency). 

I haven't worked on a package before so I must be missing something here. Can you help me out?

Edit:
I did not publish the package so it’s installed from the local package. Could that be the issue?

Edit v2:
It's fixed now. I ran `yarn pack` and installed the package from the .tgz file.

**Comment by TheFirstMe** [comment]
@jxnblk @clarkkozak All set for opening a PR. I hope the fork is ok.

**Comment by clarkkozak** [comment]
@TheFirstMe Perhaps include an example of how it works in the README and write a test for it? I haven't read it closely yet double check the [Contributions.md](https://github.com/system-ui/theme-ui/blob/master/CONTRIBUTING.md)

EDIT: And your code looks clean and readable :) I like it, yet i have no authority in this project.

**Comment by TheFirstMe** [comment]
@clarkkozak Ah forgot about the README. Just added it to the fork. It's usage is how @jxnblk mentioned in his comment above:

```js
import { jsx, ThemeProvider } from 'theme-ui'
import { MaterialUIProvider } from '@theme-ui/material-ui'
import theme from './theme'

export default props =>
  <ThemeProvider theme={theme}>
    <MaterialUIProvider>
      {props.children}
    </MaterialUIProvider>
  </ThemeProvider>
```
I have never worked with Jest before so I'm not sure about the tests. I will try making one.

**Comment by jgoux** [comment]
Hi @TheFirstMe , is there anything blocking you from submitting the PR?
The code seems almost complete. 😄

**Comment by TheFirstMe** [comment]
@jgoux I see that each package needs its tests so I'm working on that right now. I'm new to testing so it's taking some time.

**Comment by zerubeus** [comment]
Hello, I just want to say this is a very good idea, I was just trying to do something similar until I found this issue

**Comment by TheFirstMe** [comment]
@zerubeus Can you check it out and suggest any improvements or if there are any bugs?

**Comment by clarkkozak** [comment]
@jxnblk have you checked this out? It works and seems like solid code.

**Comment by TheFirstMe** [comment]
@clarkkozak Sorry I forgot about this. The tests are incomplete though. If anyone can help me with those we can open a PR.
